File: quintus3.5/generic/qplib3.5/library/aropen.c
/*  File   : aropen.c
    Author : Richard A. O'Keefe
    Updated: 12/10/98
    Purpose: Reading members of UNIX archive files.
    SeeAlso: aropen.pl, man 5 ar, man 1 ar

    Copyright (C) 1987, Quintus Computer Systems, Inc.  All rights reserved.

    This file defines
	QARopen(+Archive, +Member, -Stream) -> Errno
    which is the guts of opening an archive member.  Basically,
    we open a UNIX stdio stream to the Archive file, and search
    it sequentially for the Member.  (This is the best we can do,
    given that UNIX ar(5) files were not designed to be searched
    rapidly.)  If we find the Member, we create a Prolog stream
    for it, and return its stream code.  If not, we return a
    suitable error code.

    For comparison, you might like to look at charsio.{c,pl}
    and crypt.{c,pl} which also open user-defined streams.

    Basically, a UNIX ar(5) file has the form

	<magic><member>*
    where
	<magic> = "!<ARCH>!\n"
    and a <member> has the form
	<header><contents><pad>
    where
	<header> is an ar_hdr record describing the member,
	<contents> is the characters comprising the member,
	<pad> is 0 or 1 linefeeds, padding <contents> to an
	even number of characters.

    Two of the errors we can report are not standard unix errors,
    but are listed in library(errno) just for this operation.
    They are
	ENOTARCH == 997	Archive is not a well formed ar(5) file
	ENOTMEMB == 998	Archive is ok, but Member isn't in it.

    In System V/386 there are *two* ar(5) formats, one identified as
    "PORTAR" and one identified as "PORT5AR".  Only the one identified
    as "PORTAR" is truly portable, and that is the one we want.  Note
    that most versions of UNIX do not have PORT5AR at all.
*/
#undef	PORT5AR
#define	PORTAR  1

#include <stdio.h>
#include <ar.h>
#include <errno.h>

#include "quintus.h"
#include "qerrno.h"
#include "malloc.h"

#ifdef AIX
/* RS6000 only has its own ar format... */
#define	SARMAG	SAIAMAG
#define	ARMAG	AIAMAG
#define	ARFMAG	AIAFMAG
#define	ar_fmag	_ar_name.ar_fmag
#define	ar_name	_ar_name.ar_name
#endif


#ifndef	lint
static	char SCCSid[] = "@(#)98/12/10 aropen.c	76.1";
#endif/*lint*/

extern int strncmp(/*char^,char^,int*/);
/* [PM] 3.5 errno is often a macro these days: extern int errno; */


typedef struct ArRecord
    {
	QP_stream qpinfo;
	int	counter;	/* bytes remaining in stream */
	FILE*	archive_stream;	/* stdio stream for whole file */
	char	buffer[4];	/* mini-buffer to return chars */
    }  *ArHandle;


static int ar_read(qpstream, bufptr, sizeptr)
    QP_stream	*qpstream;
    char	**bufptr;
    size_t	*sizeptr;
    {
	register ArHandle stream = (ArHandle) qpstream;
	register int	c;

	if (--stream->counter < 0) return QP_EOF;

	if ((c = getc(stream->archive_stream)) < 0)
	{	qpstream->errnum = (errno) ? errno : QP_E_CANT_READ;
		return QP_ERROR;
	}

	stream->buffer[0] = (char) c;
	*bufptr = stream->buffer;
	*sizeptr = 1;
	return (c == '\n') ? QP_FULL : QP_PART;
    }


static int ar_close(qpstream)
    QP_stream	*qpstream;
    {
	register ArHandle stream = (ArHandle) qpstream;

	(void)fclose(stream->archive_stream);
	Free(stream);
	return QP_SUCCESS;
    }


/*  strpadeql(str, pad, len) is true when the NUL-terminated
    string str and the blank-padded counted string pad FOR len
    are equal.
*/
static int strpadeql(str, pad, len)
    register char *str;
    register char *pad;
    register int len;
    {
	register int chr;

	while ((chr = *str++))
	    if (--len < 0 || *pad++ != chr)
	        return 0;
	while (--len >= 0)
	    if (*pad++ != ' ')
		return 0;
	return 1;
    }


/*  strntol(str, len) converts the counted string str FOR len to
    an integer; it ignores everything but digits on the assumption
    that the ignored characters are padding and decoration.
*/
static long int strntol(str, len)
    register char *str;
    register int len;
    {
	register int dig;
	register long int ans = 0;

	while (--len >= 0)
	    if ((unsigned)(dig = *str++ - '0') < 10)
		ans = ans*10 + dig;
	return ans;
    }


long QARopen(Archive, Member, StreamCode)
    char *Archive;
    char *Member;
    QP_stream  **StreamCode;
    {
	register ArHandle stream;
	register FILE *archive_stream;
	char magic[SARMAG];
	long len;
	struct ar_hdr header;
	QP_stream *option;

	*StreamCode = QP_NULL_STREAM;		/* not a valid stream code */
	archive_stream = fopen(Archive, "r");
	if (!archive_stream) return errno;
	if (fread(magic, 1, SARMAG, archive_stream) != SARMAG
	 || strncmp(magic, ARMAG, SARMAG)) {
	    (void)fclose(archive_stream);
	    return ENOTARCH;
	}
	for (;;) {
	    len = fread((char*)&header, 1, sizeof header, archive_stream);
	    if (len <= 0) {
		(void)fclose(archive_stream);
		return ENOTMEMB;
	    } else
	    if (len != sizeof header
	     || strncmp(header.ar_fmag, ARFMAG, sizeof header.ar_fmag)) {
		(void)fclose(archive_stream);
		return ENOTARCH;
	    }
	    len = strntol(header.ar_size, sizeof header.ar_size);
	    if (strpadeql(Member, header.ar_name, sizeof header.ar_name)) {
		stream = Malloc(ArHandle, sizeof *stream);
		if (!stream) {
		    (void)fclose(archive_stream);
		    return ENOMEM;
		}
		stream->counter = len;
		stream->archive_stream = archive_stream;
		/* we've opened Archive, found Member, allocated
		   and initialised an ArHandle to give Prolog.
		   Now we have to talk Prolog into accepting it.
		*/

		/*  get default stream options  */
		option = &stream->qpinfo;
		QU_stream_param(Archive, QP_READ, QP_DELIM_LF, option);

		/*  modify appropriate options  */
		option->max_reclen = 1;
		option->read = ar_read;
		option->close = ar_close;
		option->seek_type = QP_SEEK_ERROR;

		/*  set Prolog system fields and register the stream  */
		QP_prepare_stream(&stream->qpinfo,
					(unsigned char *)stream->buffer);
		if (QP_register_stream(&stream->qpinfo) == QP_ERROR) {
		   (void) ar_close(&stream->qpinfo);
		   return EMFILE;
		}

		*StreamCode = &stream->qpinfo;
		return 0;
	    }
	    /* This is not the member we're looking for.  Skip it. */
	    if (len&1) len++;
	    if (fseek(archive_stream, len, 1) < 0) {
		(void)fclose(archive_stream);
		return ENOTARCH;
	    }
	}
    }

File: quintus3.5/generic/qplib3.5/library/aropen.pl
%   Package: aropen
%   Author : Richard A. O'Keefe
%   Updated: 12/10/98
%   Purpose: Reading UNIX archive entries.
%   SeeAlso: library(ar.c) /usr/include/ar.h /usr/man/man5/ar.5

%   Copyright (C) 1987, Quintus Computer Systems, Inc.  All rights reserved.

/*  This module defines one command:

	ar_open(+Archive, +Member, -Stream)

    Archive should be the name of a file which exists, can be
    opened for input, and is in UNIX ar(5) format.
    Member should be the name of a member of the archive.
    Note that UNIX archives, like the UNIX file system
    generally, are case-sensitive.  The name of an archive
    member can only be 16 characters long; this is not our
    fault, that's just the way UNIX archives are.

    ar_open/3 returns a Stream which can be used for reading
    the specified archive member, and no others.  There is a
    general restriction in user-defined streams: you cannot
    use stream_position/3 on them.

    Rewriting or appending to archive members would be possible,
    but it would require copying in general.  UNIX does not
    provide function-level tools for manipulating archives,
    only the program ar(1) {and make(1) knows about arch(memb).}

    We do not claim that this is a complete solution for people
    who want to manipulate archives from Prolog.  You should
    regard it as an example of how *YOU* can use the user-
    defined streams and C interface to provide what you really
    need.  Please let us know what you find most useful.  For
    example, do you need to be able to compile/consult out of
    archives?
*/

:- module(ar_open, [
	ar_open/3
   ]).
:- use_module(library(errno), [
	errno/2
   ]),
   use_module(library(strings), [
	string_size/2
   ]),
   use_module(library(types), [
	must_be_symbol/3,
	must_be_var/3
   ]).

sccs_id('"@(#)98/12/10 aropen.pl	76.1"').


%   ar_open(+Archive, +Member, -Stream)
%   opens a Member of a UNIX archive (see ar(1)) file Archive
%   for input, returning a Prolog Stream.

ar_open(Archive, Member, Stream) :-
	var(Stream),
	atom(Archive),
	atom(Member),
	string_size(Member, Size),
	Size > 0, Size =< 16,
	!,
	'QARopen'(Archive, Member, StreamCode, Errno),
	(   Errno =:= 0 ->
	    stream_code(Stream, StreamCode)
	;   errno(Errno, ar_open(Archive,Member,Stream))
	).
ar_open(Archive, Member, Stream) :-
	Goal = ar_open(Archive, Member, Stream),
	must_be_symbol(Archive, 1, Goal),
	must_be_symbol(Member,  2, Goal),
	must_be_var(Stream,     3, Goal),
	format(user_error,
	    '~N! Member name must be at most 16 characters~n! Goal: ~q~n',
	    [Goal]),
	fail.


foreign_file(library(system(libpl)), ['QARopen']).
foreign('QARopen', 'QARopen'(+string,+string,-address('QP_stream'),[-integer])).

:- load_foreign_executable(library(system(libpl))),
   abolish(foreign_file, 2),
   abolish(foreign, 2).

